src/rvsdg/simplify_branches.rs

```diff
@@ -0,0 +1,623 @@
+                    };
+                    let Some(val) = to_lit(&val) else {
+                        continue;
+                    };
+                    if val_analysis.data[admin_node].kills.contains(&arg) {
+                        if succ != self.exit
+                            && self.graph.neighbors(*admin_node).any(|x| x == self.exit)
+                        {
+                            // Don't remove any outgoing links to the exit node.
+                            break;
+                        }
+                        // We assign to the branched-on argument in the admin
+                        // node. See if we can fold the constant branch here.
+                        let ValueInfo::Known(lit) = val_analysis.data[admin_node].get_output(&arg)
+                        else {
+                            continue;
+                        };
+                        if lit != val {
+                            continue;
+                        }
+                        // okay, we have found a matching edge. Replace this branch
+                        // with a jump.
+                        let mut walker = self
+                            .graph
+                            .neighbors_directed(*admin_node, Direction::Outgoing)
+                            .detach();
+                        while let Some((outgoing, _)) = walker.next(&self.graph) {
+                            self.graph.remove_edge(outgoing);
+                        }
+                        self.graph.add_edge(
+                            *admin_node,
+                            succ,
+                            Branch {
+                                op: BranchOp::Jmp,
+                                pos: None,
+                            },
+                        );
+                        changed = true;
+                        // Don't run the rest of the inner loop.
+                        break;
+                    }
+                    let mut incoming_walker = self
+                        .graph
+                        .neighbors_directed(*admin_node, Direction::Incoming)
+                        .detach();
+                    while let Some((incoming, pred)) = incoming_walker.next(&self.graph) {
+                        let can_reroute = matches!(val_analysis.data[&pred].get_output(&arg), ValueInfo::Known(v) if v == val);
+                        if !can_reroute {
+                            continue;
+                        }
+
+                        let weight = self.graph.remove_edge(incoming).unwrap();
+                        // We only have to worry about `instrs` because we
+                        // checked that the footer was empty when we populated
+                        // admin_nodes. We do this because we more or less don't
+                        // use footers on our way back to bril.
+                        scratch.extend(self.graph[*admin_node].instrs.iter().cloned());
```

Comment:
Not sure I understand how `scratch` is getting used. Can `scratch` be a local definition here? What about just above this while loop instead of far up?

Reply:
scratch is being used to save repeated allocations when running the algorithm, so the higher up the better I think.

I've moved it inside `rewrite_branches` though.

Reply:
Oh I see.
I think `scratch` is a bad name, and I don't care about the repeated allocations.